### ISSUE TITLE
Define the client pool trait

### ISSUE THREAD

**Issue by fulmicoton** [quickwit-inc/quickwit#93]
The searchers receiving a root request need to dispatch the leaf work among the search servers.

We want to handle the peer discovery problem using the SWIM gossip algorithm.
Before plugging the peer discovery, please submit a documented trait describing the public API that will be consumed by the root handler.

**Comment by mosuka** [comment]
I thought about the proposed API definition for `ClientPool` trait.
Basically, I referred to the `SearchClientPool` of the `barrel`.
However, instead of doing peer discovery in the client pool, I changed it to monitor cluster events added by the SWIM protocol.
@fulmicoton @fmassot Can I have your feedback?



## API Definition Draft
### Initialize pool
The pool needs to be initialized before monitoring the events.
If there are nodes that have already joined the cluster, you need to initialize the pool with them.
```
async fn init(&self) -> ClientPoolResult<()>
```

### Watch event
Monitors the receiver of cluster events and keeps the connection pool up-to-date.
Receives events (Join, Leave, Down, etc) that occur in the cluster and updates the client pool according to those events.
```
async fn watch (&self、event_receiver: Receiver) -> ClientPoolResult<()>
```

### Assign clients
Assign the given list of jobs to the available clients.
```
async fn assign_clients(&self, jobs: Vec<Job>) -> ClientPoolResult<Vec<(SearchServiceClient, Vec<Job>)>>
```

**Comment by fmassot** [comment]
When looking at barrel code, I found it quite good to have
`async fn start(discovery: Arc<dyn Discovery>)`
so that we can switch to a basic discovery service like no discovery at all in the case where we have 1 server only.

**Comment by fulmicoton** [comment]
@fmassot but the current discovery process just gives a method to return your peer socketaddr. It requires another task to poll it and create the clients.
The polling part is not required in the case of discovery because we know precisely when new peers are available.

**Comment by fulmicoton** [comment]
Why do we need Join, Leave, Down. Isn't `Join, Leave` sufficient?

**Comment by mosuka** [comment]
@fulmicoton
There are several events defined in artillery-core. I have described these because I think they need to be handled correctly.
https://github.com/bastion-rs/artillery/blob/badac7d5519cfd6ef5078262cad7769a4d361bce/artillery-core/src/epidemic/state.rs#L31-L38

I made a mistake in describing init.
```
async fn init(&self, existing_node: Vec<(String, SocketAddress)>) -> ClientPoolResult<()>
```
I am thinking that we need to initialize the client pool bypassing the node ID and address as arguments to the init function. (The argument types are tentative.)
I think that the watch function only can only retrieve changes made to an existing cluster.
I think it is necessary to fill the client pool of the node with the information of the nodes participating in the cluster to which it connects when it starts up.

**Comment by mosuka** [comment]
@fulmicoton @fmassot 
If we watch cluster events using artillery-core, we can get the events that occurred and the cluster membership at that time, so we may not need the init function.

**Comment by fulmicoton** [comment]
> There are several events defined in artillery-core. I have described these because I think they need to be handled correctly.

We need to decouple the concern. 

There is no need for the client pool to necessarily know about all of this events. This is leaking the implementation detail of SWIM (and worse artillery-core) to the client pool. We need to only treat events that are actually meaningful. 

For instance, I don't think Payload is interesting.
Left and WentDown should be treated the same. 
WentUp and Join should probably be treated the same too.

SuspectedDown is an interesting one, because we might want to put the node aside temporarily, but avoid closing the connection.

The current API you are suggest does not offer any way to inform the cluster about a connection problem. What should we do if we fail to open a GRPC connection to a node?

**Comment by mosuka** [comment]
@fulmicoton I see. Let me think about it.

**Comment by fulmicoton** [comment]
feel free to start writing some code if it is too difficult for you to project yourself without coding.

But I want:
- no artillery-core type to be exchanged in the API
-  a clear plan on what is the spec if we cannot create a grpc connection.

**Comment by mosuka** [comment]
After conducting the PoC and confirming the operation of SWIM, the definition of the trait became very simple.

## API Definition Draft

### Assign clients
Assign the given job to the clients.
Returns a list of pair (SocketAddr, Vec<Job>)
```
async fn assign_jobs(&self, jobs: Vec<Job>) -> Result<Vec<(SocketAddr, Vec<Job>)>>;
```

**Comment by fmassot** [comment]
A first implementation has been merged in #141 
To be finished by #120
